### PR TITLE
rpc: simplify scan blocks

### PR DESCRIPTION
**Pull request by furszy**
Coming from https://github.com/bitcoin/bitcoin/pull/23549#pullrequestreview-1105712566

The current `scanblocks` flow walks-through every block in the active chain
until hits the chain tip or processes 10k blocks, then calls `lookupFilterRange`
function to obtain all filters from that particular range.

This is only done to obtain the heights range to look up the block
filters. Which is unneeded.

As `scanblocks` only lookup block filters in the active chain, we can
directly calculate the lookup range heights, by using the chain tip,
without requiring to traverse the chain block by block.

### CODE REVIEW

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -2474,8 +2474,9 @@ static RPCHelpMan scanblocks()
         } while (start_index != stop_block);
 
         ret.pushKV("from_height", start_block_height);
-        ret.pushKV("to_height", stop_block->nHeight);
+        ret.pushKV("to_height", start_index->nHeight); // start_index is always the last scanned block here
```
Comment (TheCharlatan) [inline src/rpc/blockchain.cpp:2477]: For reference,`end_range->nHeight` would be a bit easier to read, but is not guaranteed to have a value.

**Review by sedited — COMMENTED** [review]
Concept ACK

Does this require a release note for the s/filtertype/filter_type/ change?

**Comment by furszy** [comment]
Thanks for the review TheCharlatan.

> Does this require a release note for the s/filtertype/filter_type/ change?

Yeah.. I made this in December, just after #23549 got merged. Now that v25 is released, we have to add release notes for it.
Will just drop that commit.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -2383,27 +2383,30 @@ static RPCHelpMan scanblocks()
...
             }
             if (!request.params[3].isNull()) {
                 stop_block = active_chain[request.params[3].getInt<int>()];
-                if (!stop_block || stop_block->nHeight < block->nHeight) {
+                if (!stop_block || stop_block->nHeight < start_index->nHeight) {
                     throw JSONRPCError(RPC_MISC_ERROR, "Invalid stop_height");
                 }
+            } else {
+                // No stop block provided, stop at the chain tip.
+                stop_block = active_chain.Tip();
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/blockchain.cpp:2406]: yeah, thanks.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -2383,27 +2383,30 @@ static RPCHelpMan scanblocks()
...
             }
         }
-        CHECK_NONFATAL(block);
+        CHECK_NONFATAL(start_index);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/blockchain.cpp:2409]: sounds good

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (furszy) [reply src/rpc/blockchain.cpp:2477]: yeah, the user could have aborted the process at the very beginning.

maybe could change it to:
```
end_range ? end_range->nHeight : start_index->nHeight;
```
but I'm not so sure that this makes code easier to read.

**Review by sedited — APPROVED** [review]
ACK b922f6b5262884f42d7483f1e9af35650bdb53a7

**Pull request merged**